glidertest/tools.py

```diff
@@ -1100,3 +1097,161 @@ def plot_vertical_speeds_with_histograms(ds, start_prof=None, end_prof=None):
     plt.show()
 
     return fig, axs
+
+def ramsey_binavg(ds, var='VERT_CURR', zgrid=None, dz=None):
+    """
+    Calculate the bin average of vertical velocities within specified depth ranges.
+    This function computes the bin average of all vertical velocities within depth ranges,
+    accounting for the uneven vertical spacing of seaglider data in depth (but regular in time).
+    It uses the pressure data to calculate depth and then averages the vertical velocities
+    within each depth bin.
+
+    Parameters
+    ----------
+    - ds using the variables PRES and VERT_SW_SPEED
+    - zgrid (array-like, optional): Depth grid for binning. If None, a default grid is created.
+    - dz (float, optional): Interval for creating depth grid if zgrid is not provided.
+
+    Returns
+    -------
+    - meanw (array-like): Bin-averaged vertical velocities for each depth bin.
+
+    Note
+    ----
+    I know this is a non-sensical name.  We should re-name, but is based on advice from Ramsey Harcourt.
+    """
+
+    press = ds.PRES.values
+    ww = ds[var].values
+
+    # Calculate depth from pressure using gsw
+    if 'DEPTH_Z' in ds:
+        depth = ds.DEPTH_Z.values
+    else:
+        latmean = np.nanmean(ds.LATITUDE)
+        depth = gsw.z_from_p(press, lat=latmean)  # Assuming latitude is 0, adjust as necessary
+
+    if zgrid is None:
+        if dz is None:
+            dz = 5  # Default interval if neither zgrid nor dz is provided
+        zgrid = np.arange(np.floor(np.nanmin(depth)/10)*10, np.ceil(np.nanmax(depth) / 10) * 10 + 1, dz)
+
+    def findbetw(arr, bounds):
+        return np.where((arr > bounds[0]) & (arr <= bounds[1]))[0]
+
+    # Calculate bin edges from zgrid centers
+    bin_edges = np.zeros(len(zgrid) + 1)
+    bin_edges[1:-1] = (zgrid[:-1] + zgrid[1:]) / 2
+    bin_edges[0] = zgrid[0] - (zgrid[1] - zgrid[0]) / 2
+    bin_edges[-1] = zgrid[-1] + (zgrid[-1] - zgrid[-2]) / 2
+
+    meanw = np.zeros(len(zgrid))
+    NNz = np.zeros(len(zgrid))
+    w_lower = np.zeros(len(zgrid))
+    w_upper = np.zeros(len(zgrid))
+
+    # Cycle through the bins and calculate the mean vertical velocity
+    for zdo in range(len(zgrid)):
+        z1 = bin_edges[zdo]
+        z2 = bin_edges[zdo + 1]
+        ifind = findbetw(depth, [z1, z2])
+
+        CIlimits = .95 # Could be passed as a variable. 0.95 for 95% confidence intervals
+        if len(ifind):
+            meanw[zdo] = np.nanmean(ww[ifind])
+
+            # Confidence intervals
+            # Number of data points used in the mean at this depth (zgrid[zdo])
+            NNz[zdo] = len(ifind)
+            if NNz[zdo] > 1:
+                se = np.nanstd(ww[ifind]) / np.sqrt(NNz[zdo])  # Standard error
+                ci = se * t.ppf((1 + CIlimits) / 2, NNz[zdo] - 1)  # Confidence interval based on CIlimits

```

Comment:
t.ppf doesn't seem to be defined in the committed code. Does it need to be imported/calculated somewhere?

Reply:
Yep, sorry, should be stats.t.ppf.   My version had from scipy.stats import t